kornia/losses/hausdorff.py

```diff
@@ -0,0 +1,157 @@
+from typing import Callable
+
+import torch
+import torch.nn as nn
+
+
+class _HausdorffERLossBase(nn.Module):
+    """Base class for binary Hausdorff loss based on morphological erosion.
+
+    This is an Hausdorff Distance (HD) Loss that based on morphological erosion,which provided
+    a differentiable approximation of Hausdorff distance as stated in :cite:`karimi2019reducing`.
+    The code is refactored on top of `here <https://github.com/PatRyg99/HausdorffLoss/
+        blob/master/hausdorff_loss.py>`__.
+
+    Args:
+        alpha: controls the erosion rate in each iteration. Default: 2.0.
+        erosions: the number of iterations of erosion. Default: 10.
+    """
+    conv: Callable
+    avg_pool: Callable
+
+    def __init__(self, alpha: float = 2.0, erosions: int = 10) -> None:
+        super().__init__()
+        self.alpha = alpha
+        self.erosions = erosions
+        self.register_buffer("kernel", self.get_kernel())
+
+    def get_kernel(self) -> torch.Tensor:
+        raise NotImplementedError
+
+    def perform_erosion(
+        self, pred: torch.Tensor, target: torch.Tensor
+    ) -> torch.Tensor:
+        bound = (pred - target) ** 2
+
+        kernel = torch.as_tensor(self.kernel, device=pred.device, dtype=pred.dtype)
+        eroted = torch.zeros_like(bound)
+
+        for k in range(self.erosions):
+            # Same padding, assuming kernel is odd and square (cube) shaped.
+            padding = (kernel.size(-1) - 1) // 2
+            # compute convolution with kernel
+            dilation = self.conv(bound, weight=kernel, padding=padding, groups=1)
+            # apply soft thresholding at 0.5 and normalize
+            erosion = dilation - 0.5
+            erosion[erosion < 0] = 0
+
+            # image-wise differences for 2D images
+            erosion_max = self.avg_pool(erosion)
+            erosion_min = - self.avg_pool(- erosion)
+            # No normalization needed if `max - min = 0`
+            to_norm = (erosion_max.squeeze() - erosion_min.squeeze()) != 0
+            erosion[to_norm] = (erosion[to_norm] - erosion_min[to_norm]) / (
+                erosion_max[to_norm] - erosion_min[to_norm])
+
+            # save erosion and add to loss
+            eroted = eroted + erosion * (k + 1) ** self.alpha
+            bound = erosion
+
+        return eroted
+
+    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        """Compute Hausdorff loss.
+
+        Args:
+            pred: predicted tensor with a shape of :math:`(B, C, H, W)` or :math:`(B, C, D, H, W)`.
+                Each channel is as binary as: 1 -> fg, 0 -> bg.
+            target: target tensor with a shape of :math:`(B, 1, H, W)` or :math:`(B, C, D, H, W)`.
+        """
+        assert pred.shape[2:] == target.shape[2:] and pred.size(0) == target.size(0) and target.size(1) == 1, (
```

Comment:
add if / raise - same for the other assertion
https://deepsource.io/gh/kornia/kornia/run/f0e4fe5a-a07d-407b-a7c4-6e7b40f37f37/python/BAN-B101